src/librustc/ich/impls_ty.rs

```diff
@@ -26,7 +28,26 @@ for &'gcx ty::Slice<T>
     fn hash_stable<W: StableHasherResult>(&self,
                                           hcx: &mut StableHashingContext<'gcx>,
                                           hasher: &mut StableHasher<W>) {
-        (&self[..]).hash_stable(hcx, hasher);
+        thread_local! {
+            static CACHE: RefCell<FxHashMap<usize, Fingerprint>> =
+                RefCell::new(FxHashMap());
+        }
+
+        let hash = CACHE.with(|cache| {
+            let addr = &**self as *const ty::Slice<T>  as *const usize as usize;
```

Comment:
Can you check this? I tried `self as *const ty::Slice<T> as usize` but got a casting error. The suggested fix in #41300 fixed the error but I'm not really sure why it works.